include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,387 @@
+                                              state_ptr->pdu);
+                CORO_YIELD asio::async_write(state_ptr->socket.next_layer(),
+                                             asio::buffer(state_ptr->pdu),
+                                             *this);
+                auto n = asio::buffer_size(state_ptr->sequence);
+                state_ptr.invoke(error, n);
+            }
+        }
+
+        friend void* asio_handler_allocate(std::size_t size,
+                                           async_write_some_op* op) {
+            using boost::asio::asio_handler_allocate;
+            return asio_handler_allocate(
+                size, std::addressof(op->state_ptr.handler()));
+        }
+
+        friend void asio_handler_deallocate(void* p, std::size_t size,
+                                            async_write_some_op* op) {
+            using boost::asio::asio_handler_deallocate;
+            asio_handler_deallocate(p, size,
+                                    std::addressof(op->state_ptr.handler()));
+        }
+
+        friend bool asio_handler_is_continuation(async_write_some_op* op) {
+            using boost::asio::asio_handler_is_continuation;
+            return asio_handler_is_continuation(
+                std::addressof(op->state_ptr.handler()));
+        }
+
+        template <class Function>
+        friend void asio_handler_invoke(Function&& f, async_write_some_op* op) {
+            using boost::asio::asio_handler_invoke;
+            asio_handler_invoke(f, std::addressof(op->state_ptr.handler()));
+        }
+
+        struct state {
+            template <typename DeducedHandler>
+            state(DeducedHandler&&, secure_socket& socket,
+                  const ConstBufferSequence& sequence)
+                : socket{socket}, sequence{sequence} {}
+
+            secure_socket& socket;
+            ConstBufferSequence sequence;
+            std::vector<gsl::byte> pdu;
+        };
+
+        beast::handler_ptr<state, Handler> state_ptr;
+    };
+
+    template <typename MutableBufferSequence, typename Handler>
+    struct async_read_some_op : public asio::coroutine {
+        template <typename DeducedHandler>
+        async_read_some_op(secure_socket& socket,
+                           const MutableBufferSequence& sequence,
+                           DeducedHandler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<DeducedHandler>(handler)} {}
+        secure_socket& socket;
+        MutableBufferSequence sequence;
+        Handler handler;
+
+        size_t transferred = 0;
+        boost::endian::little_int16_t payload_length = 0;
+        boost::optional<beast::flat_buffer::mutable_buffers_type> raw_sequence;
+
+        void operator()(beast::error_code error, size_t n) {
+            CORO_REENTER(*this) {
+                while (transferred < sizeof(payload_length) + payload_length +
```

Comment:
Need to attempt length extraction before the loop, because the raw buffer might contain unparsed PDUs.